WHY_DI.md

```diff
@@ -0,0 +1,27 @@
+
+Now let's step back and analyze our code. In its current state, the `PhotosViewController` and `PhotosService` are tightly coupled. This leaves us with a few issues:
+1. We cannot change `PhotosService` without having to also change `PhotosViewController`. This may seem fine with just two classes, but in a real-world scenario with hundreds of classes this would significantly slowdown our app iteration.
+2. We cannot switch out the `PhotosService` class without having to change `PhotosViewController`. Let's imagine we have a better `PhotosServiceV2` class we now want our view controller to use, we'll have to dig into the implementation of `PhotosViewController` to make changes.
+3. We cannot unit test `PhotosViewController` without also invoking the `PhotosService` implementation.
+4. We cannot develop `PhotosViewController` independently and concurrently with the `PhotosService`. This may not seem like a big deal with our extremely simple app, in a real-world team, our engineers would be blocked constantly.
+
+Let's apply the dependency injection pattern to our app. With DI, we will have a third class, in Needle's terms a `Component` class, that instantiates the `PhotosService` and pass it into the `PhotosViewController` as a protocol. Let's call this protocol `PhotosServicing`. Now our `PhotosViewController` no longer knows anything about the concrete implementation of `PhotosService`. It simply uses the passed in `PhotosServicing` protocol to perform its logic.
+
+With DI applied, let's revisit the issues we had before:
+1. We can freely change the implementation of `PhotosService` without affecting our `PhotosViewController`.
+2. We can simply update our DI `Component` class to instantiate and pass `PhotosServiceV2` into `PhotosViewController`, as long as the implementaiton still conforms to the `PhotosServicing` protocol. This allows us to freely switch implementations of the photos service without having to change anything in the view controller.
+3. We can properly unit test `PhotosViewController` by injecting, aka passing in, a mock `PhotosServicing` object.
+4. As soon as the `PhotosServicing` protocol is defined, we can independently and concurrently develop `PhotosService` and `PhotosViewController` implementations.
+
+## Dependency injection terminologies
+
+Before moving on, let's define some terminologies that are frequently used with the dependency injection pattern. In our example app above, the `PhotosService` is typically referred to as a "dependency". Our `PhotosViewController` class is sometimes referred to as the "dependent" or "consumer". The act of passing an instance of `PhotosServicing` into the `PhotosViewController` is called  "inject". In summary, our simple DI setup injects the `PhotosServicing` dependency into the consumer `PhotosViewController`. 
```

Comment:
"terminologies that are" -> "terminology that is"

Reply:
👍